### PR TITLE
[@types/chrome] Define chrome.tabGroups API available in Chrome 89

### PR DESCRIPTION
**Pull request by userTim**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://developer.chrome.com/docs/extensions/reference/tabGroups>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by zwacky** [comment]
Hey @usertim, this PR is missing the new prop `tabs.groupId: number`.

if the tab isn't in a tabGroup, it will hold the value `-1`.

**Comment by usertim** [comment]
Hey @zwacky 
Thanks for pointing out!
Extended chrome.tabs interface with missing props and method related to tab groups.

File: types/chrome/index.d.ts [hunk 1]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+     */
+     export function query(queryInfo: QueryInfo, callback: (result: TabGroup[]) => void): void;
+
+    /**
+     * Modifies the properties of a group. Properties that are not specified in updateProperties are not modified.
+     * @param groupId The ID of the group to modify.
+     * @param updateProperties The properties to update.
+     * @param callback Called with an updated tab group as parameter.
+     */
+    export function update(groupId: string, updateProperties: UpdateProperties, callback?: (group: TabGroup) => void): void;
```
Comment (zwacky) [inline types/chrome/index.d.ts:3894]: `update` will only work, if the groupId is a number, just ran into this myself.

Comment (zwacky) [reply types/chrome/index.d.ts]: thank you for your swift changes, @usertim.

**Review by sandersn — CHANGES_REQUESTED** [review]
Missing some indentation. You can try running prettier on this file, but if it's never been run on it before, it would cause less churn to fix it manually.

File: types/chrome/index.d.ts [hunk 2]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+     * Retrieves details about the specified group.
+     * @param groupId The ID of the tab group.
+     * @param callback Called with the retrieved tab group.
+     */
+    export function get(groupId: number, callback: (group: TabGroup) => void): void;
+
+    /**
+     * Moves the group and all its tabs within its window, or to a new window.
+     * @param groupId The ID of the group to move.
+     * @param moveProperties Move properties. Specify tab index, windowId
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:3876]:
```suggestion
     * @param moveProperties Information on how to move the group.
```

File: types/chrome/index.d.ts [hunk 3]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+     * @param callback Called with details about the moved group.
+    */
+    export function move(groupId: number, moveProperties: MoveProperties, callback: (group: TabGroup) => void): void;
+
+    /**
+     * Queries all groups that have the specified properties, or all groups if no properties are specified.
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:3882]:
```suggestion
     * Gets all groups that have the specified properties, or all groups if no properties are specified.
```

File: types/chrome/index.d.ts [hunk 4]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+     * @param queryInfo Object with search parameters.
+     * @param callback Called with retrieved tab groups
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:3884]:
```suggestion
     * @param callback Optional.
     * Parameter result: Details about the groups matching the query.
```

File: types/chrome/index.d.ts [hunk 5]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+     */
+     export function query(queryInfo: QueryInfo, callback: (result: TabGroup[]) => void): void;
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:3886]:
```suggestion
     export function query(queryInfo: QueryInfo, callback?: (result: TabGroup[]) => void): void;
```

File: types/chrome/index.d.ts [hunk 6]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+
+    /**
+     * Modifies the properties of a group. Properties that are not specified in updateProperties are not modified.
+     * @param groupId The ID of the group to modify.
+     * @param updateProperties The properties to update.
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:3891]:
```suggestion
     * @param updateProperties Information on how to update the group.
```

File: types/chrome/index.d.ts [hunk 7]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+     * @param callback Called with an updated tab group as parameter.
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:3892]:
```suggestion
     * @param callback Optional.
     * Parameter group: Details about the updated group.
```

File: types/chrome/index.d.ts [hunk 8]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+     */
+    export function update(groupId: number, updateProperties: UpdateProperties, callback?: (group: TabGroup) => void): void;
+
+    export interface TabGroupCreatedEvent extends chrome.events.Event<(group: TabGroup) => void> { }
+	export interface TabGroupMovedEvent extends chrome.events.Event<(group: TabGroup) => void> { }
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:3897]:
```suggestion
    export interface TabGroupMovedEvent extends chrome.events.Event<(group: TabGroup) => void> { }
```

File: types/chrome/index.d.ts [hunk 9]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+	export interface TabGroupRemovedEvent extends chrome.events.Event<(group: TabGroup) => void> { }
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:3898]:
```suggestion
    export interface TabGroupRemovedEvent extends chrome.events.Event<(group: TabGroup) => void> { }
```

File: types/chrome/index.d.ts [hunk 10]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+	export interface TabGroupUpdated extends chrome.events.Event<(group: TabGroup) => void> { }
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:3899]:
```suggestion
    export interface TabGroupUpdated extends chrome.events.Event<(group: TabGroup) => void> { }
```

File: types/chrome/index.d.ts [hunk 11]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+
+    /**
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:3901]: Nit: Make these single line comments

File: types/chrome/index.d.ts [hunk 12]
```diff
@@ -7721,6 +7830,11 @@ declare namespace chrome.tabs {
          * @since Chrome 31.
          */
         sessionId?: string;
+		/**
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:7833]:
```suggestion
        /**
```

File: types/chrome/index.d.ts [hunk 13]
```diff
@@ -7721,6 +7830,11 @@ declare namespace chrome.tabs {
...
+		 * The ID of the group that the tab belongs to.
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:7834]:
```suggestion
        * The ID of the group that the tab belongs to.
```

File: types/chrome/index.d.ts [hunk 14]
```diff
@@ -7721,6 +7830,11 @@ declare namespace chrome.tabs {
...
+		 * @since Chrome 88
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:7835]:
```suggestion
        * @since Chrome 88
```

File: types/chrome/index.d.ts [hunk 15]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
     export var onSendError: GcmErrorEvent;
 }
 
+////////////////////
+// Tab Groups
+////////////////////
+/**
+ * Use the chrome.tabGroups API to interact with the browser's tab grouping system. You can use this API to modify and rearrange tab groups in the browser. To group and ungroup tabs, or to query what tabs are in groups, use the chrome.tabs API.
+ * Permissions:  "tabGroups"
+ * @since Chrome 89.
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:3817]:
```suggestion
 * @since Chrome 89. Manifest V3 and above.
```

File: types/chrome/index.d.ts [hunk 16]
```diff
@@ -7721,6 +7830,11 @@ declare namespace chrome.tabs {
          * @since Chrome 31.
          */
         sessionId?: string;
+		/**
+		 * The ID of the group that the tab belongs to.
+		 * @since Chrome 88
+		 */
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:7836]:
```suggestion
        */
```

File: types/chrome/index.d.ts [hunk 17]
```diff
@@ -7721,6 +7830,11 @@ declare namespace chrome.tabs {
...
+		groupId: number;
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:7837]:
```suggestion
        groupId: number;
```

File: types/chrome/index.d.ts [hunk 18]
```diff
@@ -7904,6 +8018,18 @@ declare namespace chrome.tabs {
         frameId?: number;
     }
 
+	export interface GroupOptions {
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:8021]:
```suggestion
    export interface GroupOptions {
```

File: types/chrome/index.d.ts [hunk 19]
```diff
@@ -7904,6 +8018,18 @@ declare namespace chrome.tabs {
...
+		/** Configurations for creating a group. Cannot be used if groupId is already specified. */
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:8022]:
```suggestion
        /** Configurations for creating a group. Cannot be used if groupId is already specified. */
```

File: types/chrome/index.d.ts [hunk 20]
```diff
@@ -7904,6 +8018,18 @@ declare namespace chrome.tabs {
...
+		createProperties?: {
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:8023]:
```suggestion
        createProperties?: {
```

File: types/chrome/index.d.ts [hunk 21]
```diff
@@ -7904,6 +8018,18 @@ declare namespace chrome.tabs {
...
+			/** The window of the new group. Defaults to the current window. */
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:8024]:
```suggestion
            /** The window of the new group. Defaults to the current window. */
```

File: types/chrome/index.d.ts [hunk 22]
```diff
@@ -7904,6 +8018,18 @@ declare namespace chrome.tabs {
...
+			windowId?: number
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:8025]:
```suggestion
            windowId?: number
```

File: types/chrome/index.d.ts [hunk 23]
```diff
@@ -7904,6 +8018,18 @@ declare namespace chrome.tabs {
...
+		},
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:8026]:
```suggestion
        },
```

File: types/chrome/index.d.ts [hunk 24]
```diff
@@ -7904,6 +8018,18 @@ declare namespace chrome.tabs {
...
+		/** The ID of the group to add the tabs to. If not specified, a new group will be created. */
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:8027]:
```suggestion
        /** The ID of the group to add the tabs to. If not specified, a new group will be created. */
```

File: types/chrome/index.d.ts [hunk 25]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
     export var onSendError: GcmErrorEvent;
 }
 
+////////////////////
```
Comment (ThatGuyRussell) [inline types/chrome/index.d.ts:3811]: Please add tests. Reference: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51891/files#diff-1ac7b611fa281ad873e71ccb732137409dbeb934e1507a564cf1f108831540b4

**Review by ThatGuyRussell — CHANGES_REQUESTED** [review]

File: types/chrome/index.d.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (usertim) [reply types/chrome/index.d.ts]: Moved

File: types/chrome/index.d.ts [hunk 26]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+ * Use the chrome.tabGroups API to interact with the browser's tab grouping system. You can use this API to modify and rearrange tab groups in the browser. To group and ungroup tabs, or to query what tabs are in groups, use the chrome.tabs API.
+ * Permissions:  "tabGroups"
+ * @since Chrome 89.
+ */
+ declare namespace chrome.tabGroups {
+
+    /** An ID that represents the absence of a group. */
+    export var TAB_GROUP_ID_NONE: number;
+
+	type ColorEnum = "grey" | "blue" | "red" | "yellow" | "green" | "pink" | "purple" | "cyan";
```
[the remark this answers is not in the document]
Comment (usertim) [reply types/chrome/index.d.ts]: Fixed

File: types/chrome/index.d.ts [hunk 27]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+
+	interface TabGroup {
```
[the remark this answers is not in the document]
Comment (usertim) [reply types/chrome/index.d.ts]: Fixed all indentations

File: types/chrome/index.d.ts [hunk 28]
```diff
@@ -3807,6 +3808,114 @@ declare namespace chrome.gcm {
...
+        /** The ID of the parent window, or chrome.windows.WINDOW_ID_CURRENT for the current window. */
+        windowId?: number;
+    }
+
+    /**
+     * Retrieves details about the specified group.
+     * @param groupId The ID of the tab group.
+     * @param callback Called with the retrieved tab group.
+     */
+    export function get(groupId: number, callback: (group: TabGroup) => void): void;
```
[the remark this answers is not in the document]
Comment (usertim) [reply types/chrome/index.d.ts]: As of 89.0.4389.90 build `.get` and `.query` method callbacks are not optional.
And while it is not stated in docs `.update` and `.move` methods will work without callback parameter just fine.

File: types/chrome/index.d.ts (hunk 25 above)
Comment (usertim) [reply types/chrome/index.d.ts]: Added

**Review by ThatGuyRussell — APPROVED** [review]

**Review by andrewbranch — APPROVED** [review]

**Comment by usertim** [comment]
Ready to merge

**Pull request merged**